core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -245,15 +249,17 @@ class TransactionStateManager(brokerId: Int,
       else if (leavingPartitions.exists(_.txnPartitionId == partitionId))
         Left(Errors.NOT_COORDINATOR)
       else {
+        var isNewMetadata = false
         transactionMetadataCache.get(partitionId) match {
           case Some(cacheEntry) =>
             val txnMetadata = Option(cacheEntry.metadataPerTransactionalId.get(transactionalId)).orElse {
+              isNewMetadata = true
```

Comment:
There seems to be a race condition here. The `putIfNotExists` below ensures that only one thread can succeed in inserting the metadata, but it seems `isNewMetadata` could be true for concurrent threads. This may be fixed if we conclude that it is safe to get rid of the `isNewMetadata` flag as suggested previously.